Feat: Split the Drawer component

uskayyyyy opened the pull request:
<!--

Title Naming Conventions:

- Start with a capitalized prefix followed by an imperative form of a verb, just like a commit message.
- Refer to the "Commit Convention" for the prefix:
  https://github.com/genesis-tech-tribe/nishiki-frontend/blob/develop/docs/CONTRIBUTING.md#commit-convention.

Example:
- Fix: Resolve memory leak in data processing module
- Feat: Add two-factor authentication
- Refactor: Improve state management in React components
- Docs: Add Windows setup instructions

-->

## Overview

<!-- Provide a brief description of the changes introduced by this PR. -->

This pull request introduces two new components: `BaseDrawer` and `PrimaryDrawer`. The purpose of this change is to demonstrate how to follow the [Open-Closed Principle of the SOLID principles](https://www.freecodecamp.org/news/open-closed-principle-solid-architecture-concept-explained/) when creating multiple variations of drawer components. Please note that the existing drawer components (`Drawer.tsx`, `DrawerUsageExample.tsx`, `SecondaryDrawer.tsx`, and `SecondaryDrawerUsageExample.tsx`) have intentionally been kept for reference.

## Changes

<!-- List the changes -->
<!-- Delete this section if not needed -->

- Created the `BaseDrawer` component.
- Created the `PrimaryBottomDrawer` component.
- Created the `PrimaryBottomDrawerUsageExampel`.
- Put all the newly created files in the `Drawers/` directory.

## Review points

<!-- List the points you would like to be reviewed in detail and the points you are not confident about. -->
<!-- Delete this section if not needed -->

- 

## Screen Captures

<!-- Add screenshots or screen recordings of the changes -->
<!-- Necessary for UI changes -->
<!-- Delete this section if not needed -->

## Note

<!-- Write any note or comment. You can share your thoughts or ideas. -->
<!-- Delete this section if not needed -->

- The next task will be #188 
- The implementation of this PR is an example of creating a customized drawer for various needs.
- Thus, I renamed `PrimaryBottomDrawer` to `ExampleDrawer` since we no longer need the drawer with a primary color background.

## Assignee Checklist:

<!-- You can tick the checkboxes if not applicable -->

- [x] The naming convention of the PR title is correct (See the comment at the top of this template)
- [x] The base branch is correct (See: [Types of Branches](https://github.com/genesis-tech-tribe/nishiki-frontend/blob/develop/docs/CONTRIBUTING.md#types-of-branches))
- [x] The branch name follows the [Branch Naming Conventions](https://github.com/genesis-tech-tribe/nishiki-frontend/blob/develop/docs/CONTRIBUTING.md#branch-naming-conventions)
- [x] The correct assignees and reviewers have been designated for this PR
- [x] The coding style follows the [Coding Style Guide](https://github.com/genesis-tech-tribe/nishiki-frontend/blob/develop/docs/STYLEGUIDE.md)
- [x] All the related issues are associated with this PR
- [x] All criteria in the associated issues are met (please tick the checkboxes)
- [x] My changes do not generate new warnings or errors (especially in the console)

## Reviewer Checklist:

<!-- You can tick the checkboxes if not applicable -->

- [ ] Double-check the "Assignee Checklist"
- [ ] The code follows the generic best practices and our [Coding Style Guide](https://github.com/genesis-tech-tribe/nishiki-frontend/blob/develop/docs/STYLEGUIDE.md)
- [ ] The code is well-commented and easy to understand
- [ ] The UI changes accurately reflect the provided design

uskayyyyy commented on src/components/ui/Drawers/BaseDrawer.tsx:
```diff
@@ -0,0 +1,163 @@
+        'inset-y-0 right-0 h-full w-5/6 max-w-sm',
+        'data-[state=open]:animate-slideInFromRight data-[state=closed]:animate-slideOutToRight',
+      ),
+    },
```
I decided not to add a `color` or `variant` property, which would look like the following code because the CSS `border-color` property is not inherited by children elements by the CSS3 default feature.

```tsx
variant: {
  white: 'bg-white border-gray-light',
  primary: 'bg-primary-lightest border-primary-light',
},
```

kotaaaa reviewed (approved):
LGTM:)

kanta1207 commented on src/components/ui/Drawers/PrimaryBottomDrawer.tsx:
```diff
@@ -0,0 +1,101 @@
+const PrimaryBottomDrawerOverlay = BaseDrawerOverlay;
+
+interface IPrimaryBottomDrawerContentProps
+  extends ComponentPropsWithoutRef<typeof BaseDrawerContent> {}
```
### [Question & Optional suggestion]
Did you decide not to omit the `side` prop from this `PrimaryBottomDrawer`? 
 Since it's fixed to be used as Bottom drawer, I thought we were going to omit it like this ⏬ 
 ```tsx
 interface IPrimaryBottomDrawerContentProps
  extends Omit<ComponentPropsWithoutRef<typeof BaseDrawerContent>, 'side'> {}
  ```
 With current implementation, I suppose we can still pass `right` to `side` props of `PrimaryBottomDrawer` component.

uskayyyyy replied:
I am sorry I forgot it. Please allow me some time to update it.

uskayyyyy replied:
@kanta1207 

I fixed in this commit: [`cd19587` (#186)](https://github.com/genesis-tech-tribe/nishiki-frontend/pull/186/commits/cd19587537e7345bcfa90a78f44e0ca8af765c5e). Please have a check.

kanta1207 reviewed (commented):
Thank you for the implementation!!
Overall, it now seems to be OPEN for extension, but CLOSED  for fixes!

Please just take a look at my [Question & Optional suggestion] on the usage of `Omit` in `PrimaryBottomDrawer` component 🙏

uskayyyyy commented:
@kotaaaa @kanta1207 @ShoeheyOt 

Hi guys, please let me share my thoughts on how to handle this PR. As per our earlier verbal discussion, we have decided that we no longer need the `PrimaryBottomDrawer`. However, I think it would still be good to merge this PR, maybe along with [@ShoeheyOt 's PR](https://github.com/genesis-tech-tribe/nishiki-frontend/pull/177), to preserve the `BaseDrawer` and provide an example of how to implement a customized drawer for each need by inheriting `BaseDrawer`. We just need to make sure to delete the `SecondaryDrawer` when merging into the `develop` branch.

Therefore, I renamed `PrimaryBottomDrawer` to `ExampleDrawer` and added some comments to indicate that it's an example file.

To know how I updated after the verbal discussion, please take a look at the commits after [`cd19587` (#186)](https://github.com/genesis-tech-tribe/nishiki-frontend/pull/186/commits/cd19587537e7345bcfa90a78f44e0ca8af765c5e)

I'd also love to hear your opinions.

kanta1207 commented:
> @kotaaaa @kanta1207 @ShoeheyOt
> 
> Hi guys, please let me share my thoughts on how to handle this PR. As per our earlier verbal discussion, we have decided that we no longer need the `PrimaryBottomDrawer`. However, I think it would still be good to merge this PR, maybe along with [@ShoeheyOt 's PR](https://github.com/genesis-tech-tribe/nishiki-frontend/pull/177), to preserve the `BaseDrawer` and provide an example of how to implement a customized drawer for each need by inheriting `BaseDrawer`. We just need to make sure to delete the `SecondaryDrawer` when merging into the `develop` branch.
> 
> Therefore, I renamed `PrimaryBottomDrawer` to `ExampleDrawer` and added some comments to indicate that it's an example file.
> 
> To know how I updated after the verbal discussion, please take a look at the commits after [`cd19587` (#186)](https://github.com/genesis-tech-tribe/nishiki-frontend/pull/186/commits/cd19587537e7345bcfa90a78f44e0ca8af765c5e)
> 
> I'd also love to hear your opinions.

Thank you for sharing & summarizing your thoughts on this and I totally agree with you!
I especially liked your idea on renaming `PrimaryBottomDrawer` to be `Example`, that's really a nice way to keep this work and Shu's effort for future!

uskayyyyy commented:
@kanta1207 @kotaaaa @ShoeheyOt 
Thank you for your feedback :)
I will merge it now!
